File: platform/analysis-api/src/com/intellij/codeInsight/completion/kind/CompletionKind.java
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.codeInsight.completion.kind;

import com.intellij.codeInsight.completion.CompletionSession;
import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.openapi.util.Key;

public abstract class CompletionKind {
  public static final Key<CompletionKind> LOOKUP_ELEMENT_COMPLETION_KIND = Key.create("lookup element completion kind");

  private final String name;
  private boolean alreadyFilled = false;

  public CompletionKind(String name) { this.name = name; }

  protected abstract void fillKindVariants();

  public abstract boolean isApplicable();

  public String getName() { return name; }

  public void fillKindVariantsOnce(CompletionSession session) {
    if (alreadyFilled) {
      return;
    }
    session.getResult().setCurrentCompletionKind(this);
    try {
      fillKindVariants();
    }
    finally {
      alreadyFilled = true;
      session.getResult().resetCurrentCompletionKind();
    }
  }

  public void putKindInfoRequireEmpty(LookupElement element) {
    if (element.getUserData(LOOKUP_ELEMENT_COMPLETION_KIND) != null) {
      throw new IllegalStateException("For element: '%s' LOOKUP_ELEMENT_COMPLETION_KIND must be empty");
    }
    element.putUserData(LOOKUP_ELEMENT_COMPLETION_KIND, this);
  }
}

File: java/java-impl/src/com/intellij/codeInsight/completion/kind/CompletionKind.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.codeInsight.completion.kind

import com.intellij.codeInsight.completion.kind.state.Flag
import com.intellij.codeInsight.completion.kind.state.FlagWithSupplier
import com.intellij.codeInsight.completion.kind.state.LatestValueTakingFlag


abstract class CompletionKind(val name: String) {
  private var alreadyFilled = false

  protected abstract fun fillKindVariants()
  abstract val isApplicable: Flag

  fun fillKindVariantsOnce() {
    if (alreadyFilled) {
      return
    }
    fillKindVariants()
    alreadyFilled = true
  }

  companion object {
    private class EmptyVariantFiller : Runnable {
      override fun run() = throw UnsupportedOperationException()
    }

    @JvmStatic
    fun withCompletionDecision(
      name: String,
      isApplicable: Flag,
      doFillVariants: Runnable,
    ): CompletionKindWithMutableFiller {

      class CompletionKindWithDynamicCompletionDecision : CompletionKindWithMutableFiller(name, isApplicable, doFillVariants) {
        override fun fillKindVariants() = variantFiller.run()
        fun setFiller(newVariantFiller: Runnable) {
          variantFiller = newVariantFiller
        }
      }

      return CompletionKindWithDynamicCompletionDecision()
    }

    @JvmStatic
    fun withCompletionDecision(
      name: String,
      isKindApplicable: Flag
    ) = withCompletionDecision(name, isKindApplicable, EmptyVariantFiller());

    @JvmStatic
    fun withCompletionDecision(
      name: String,
      isKindApplicable: () -> Boolean,
      doFillVariants: Runnable,
    ) = withCompletionDecision(name, FlagWithSupplier(isKindApplicable).withLazySupplier(), doFillVariants);

    @JvmStatic
    fun withFillFunction(
      name: String,
      doFillVariants: Runnable
    ): CompletionKindWithMutableFiller = withCompletionDecision(name,
                                                                LatestValueTakingFlag(true),
                                                                doFillVariants)

    @JvmStatic
    fun withEmptyFillFunction(
      name: String
    ) = withFillFunction(name, EmptyVariantFiller())
  }
}

open class CompletionKindWithMutableFiller(
  name: String,
  override val isApplicable: Flag,
  var variantFiller: Runnable
) : CompletionKind(name) {
  override fun fillKindVariants() = variantFiller.run()
}
